File: waqt_ci4/app/Controllers/Api/AuthController.php
<?php

namespace App\Controllers\Api;

use App\Controllers\BaseController;
use App\Models\UserModel;
use CodeIgniter\API\ResponseTrait;

class AuthController extends BaseController
{
    use ResponseTrait;

    public function register()
    {
        $json = $this->request->getJSON(true);
        $username = isset($json['username']) ? trim($json['username']) : null;
        $password = isset($json['password']) ? trim($json['password']) : null;

        if (!$username || !$password) {
            return $this->respond([
                'status'  => 'error',
                'message' => 'Username and password required'
            ], 400);
        }

        $userModel = new UserModel();
        
        // Check if username already exists
        if ($userModel->where('username', $username)->first()) {
            return $this->respond([
                'status'  => 'error',
                'message' => 'Username already exists'
            ], 400);
        }

        $passwordHash = password_hash($password, PASSWORD_BCRYPT);
        $sessionToken = bin2hex(random_bytes(16));

        $userData = [
            'username'      => $username,
            'password_hash' => $passwordHash,
            'session_token' => $sessionToken
        ];

        if ($userModel->insert($userData)) {
            return $this->respond([
                'status'   => 'success',
                'token'    => $sessionToken,
                'username' => $username
            ], 201);
        }

        return $this->respond([
            'status'  => 'error',
            'message' => 'Registration failed'
        ], 500);
    }

    public function login()
    {
        $json = $this->request->getJSON(true);
        $username = isset($json['username']) ? trim($json['username']) : null;
        $password = isset($json['password']) ? trim($json['password']) : null;

        if (!$username || !$password) {
            return $this->respond([
                'status'  => 'error',
                'message' => 'Username and password required'
            ], 400);
        }

        $userModel = new UserModel();
        $user = $userModel->where('username', $username)->first();

        if (!$user || !password_verify($password, $user['password_hash'])) {
            return $this->respond([
                'status'  => 'error',
                'message' => 'Invalid credentials'
            ], 401);
        }

        $sessionToken = bin2hex(random_bytes(16));
        $userModel->update($user['id'], ['session_token' => $sessionToken]);

        return $this->respond([
            'status'   => 'success',
            'token'    => $sessionToken,
            'username' => $username
        ], 200);
    }

    public function updateProfile()
    {
        $userId = $this->request->getHeaderLine('X-User-ID');
        if (!$userId) {
            return $this->respond([
                'status'  => 'error',
                'message' => 'Unauthorized'
            ], 401);
        }

        $json = $this->request->getJSON(true);
        $username = isset($json['username']) ? trim($json['username']) : null;
        $password = isset($json['password']) ? trim($json['password']) : null;

        if (!$username && !$password) {
            return $this->respond([
                'status'  => 'error',
                'message' => 'Username or password required to update'
            ], 400);
        }

        $userModel = new UserModel();
        $updateData = [];

        if ($username) {
            // Check if username already exists for another user
            $existing = $userModel->where('username', $username)->first();
            if ($existing && $existing['id'] != $userId) {
                return $this->respond([
                    'status'  => 'error',
                    'message' => 'Username already exists'
                ], 400);
            }
            $updateData['username'] = $username;
        }

        if ($password) {
            $updateData['password_hash'] = password_hash($password, PASSWORD_BCRYPT);
        }

        if ($userModel->update($userId, $updateData)) {
            $response = [
                'status'  => 'success',
                'message' => 'Profile updated successfully'
            ];
            if ($username) {
                $response['username'] = $username;
            }
            return $this->respond($response, 200);
        }

        return $this->respond([
            'status'  => 'error',
            'message' => 'Failed to update profile'
        ], 500);
    }
}

File: waqt_springboot/src/main/java/com/waqt/api/controller/AuthController.java
package com.waqt.api.controller;

import com.waqt.api.entity.User;
import com.waqt.api.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    @Autowired
    private UserRepository userRepository;

    private final BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();

    @PostMapping("/register")
    public ResponseEntity<Map<String, Object>> register(@RequestBody Map<String, String> request) {
        String username = request.get("username");
        String password = request.get("password");

        if (username == null || username.trim().isEmpty() || password == null || password.trim().isEmpty()) {
            Map<String, Object> response = new HashMap<>();
            response.put("status", "error");
            response.put("message", "Username and password required");
            return ResponseEntity.badRequest().body(response);
        }

        username = username.trim();
        password = password.trim();

        if (userRepository.findByUsername(username).isPresent()) {
            Map<String, Object> response = new HashMap<>();
            response.put("status", "error");
            response.put("message", "Username already exists");
            return ResponseEntity.badRequest().body(response);
        }

        String passwordHash = passwordEncoder.encode(password);
        String sessionToken = UUID.randomUUID().toString().replace("-", "");

        User user = User.builder()
                .username(username)
                .passwordHash(passwordHash)
                .sessionToken(sessionToken)
                .build();

        userRepository.save(user);

        Map<String, Object> response = new HashMap<>();
        response.put("status", "success");
        response.put("token", sessionToken);
        response.put("username", username);

        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @PostMapping("/login")
    public ResponseEntity<Map<String, Object>> login(@RequestBody Map<String, String> request) {
        String username = request.get("username");
        String password = request.get("password");

        if (username == null || username.trim().isEmpty() || password == null || password.trim().isEmpty()) {
            Map<String, Object> response = new HashMap<>();
            response.put("status", "error");
            response.put("message", "Username and password required");
            return ResponseEntity.badRequest().body(response);
        }

        username = username.trim();
        password = password.trim();

        Optional<User> userOpt = userRepository.findByUsername(username);

        if (userOpt.isEmpty() || !passwordEncoder.matches(password, userOpt.get().getPasswordHash())) {
            Map<String, Object> response = new HashMap<>();
            response.put("status", "error");
            response.put("message", "Invalid credentials");
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(response);
        }

        User user = userOpt.get();
        String sessionToken = UUID.randomUUID().toString().replace("-", "");
        user.setSessionToken(sessionToken);
        userRepository.save(user);

        Map<String, Object> response = new HashMap<>();
        response.put("status", "success");
        response.put("token", sessionToken);
        response.put("username", username);

        return ResponseEntity.ok(response);
    }

    @RequestMapping(value = "/update", method = {RequestMethod.POST, RequestMethod.PUT})
    public ResponseEntity<Map<String, Object>> updateProfile(
            @RequestHeader(value = "X-User-ID", required = false) String userIdHeader,
            @RequestBody Map<String, String> request) {

        if (userIdHeader == null) {
            Map<String, Object> response = new HashMap<>();
            response.put("status", "error");
            response.put("message", "Unauthorized");
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(response);
        }

        Long userId = Long.parseLong(userIdHeader);
        String username = request.get("username");
        String password = request.get("password");

        if ((username == null || username.trim().isEmpty()) && (password == null || password.trim().isEmpty())) {
            Map<String, Object> response = new HashMap<>();
            response.put("status", "error");
            response.put("message", "Username or password required to update");
            return ResponseEntity.badRequest().body(response);
        }

        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            Map<String, Object> response = new HashMap<>();
            response.put("status", "error");
            response.put("message", "User not found");
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
        }

        User user = userOpt.get();

        if (username != null && !username.trim().isEmpty()) {
            username = username.trim();
            Optional<User> existing = userRepository.findByUsername(username);
            if (existing.isPresent() && !existing.get().getId().equals(userId)) {
                Map<String, Object> response = new HashMap<>();
                response.put("status", "error");
                response.put("message", "Username already exists");
                return ResponseEntity.badRequest().body(response);
            }
            user.setUsername(username);
        }

        if (password != null && !password.trim().isEmpty()) {
            user.setPasswordHash(passwordEncoder.encode(password.trim()));
        }

        userRepository.save(user);

        Map<String, Object> response = new HashMap<>();
        response.put("status", "success");
        response.put("message", "Profile updated successfully");
        response.put("username", user.getUsername());

        return ResponseEntity.ok(response);
    }
}
